docs/boundaries.md

```diff
@@ -0,0 +1,131 @@
+
+In general, Ion text values can appear adjacent to each other when the boundary between the values is unambiguous. The boundaries between quoted symbols, double-quoted strings, triple-quoted strings, lobs, structs, lists, and S-expressions are unambiguous, so these elements can appear adjacent to one another in Ion text:
+
+```ion
+{% raw %}
+// A valid Ion data stream
+(1)[2]{a:struct}'hello'"world"{{ SSBsb3ZlIElvbiE= }}{{ "Same with clob" }}'annotations too!'::123
+{% endraw %}
+```
+
+Additionally, any value can appear immediately *after* a value of these types, including numeric values, keywords, and identifiers.
+
+Container separators delimit values unambiguously, so any value can also appear adjacent to a container boundary: 
+
+```ion
+// Some valid, compact containers
+[1,2,"three"]
+{name:"Austin",message:"Hello world!"}
+```
+
+Indentifiers (unquoted symbols) and reserved keywords (`null` and its typed variants, `true`, `false`, and `nan`) are terminated by the first non-identifier character encountered. The identifier characters are ASCII letters, digits, or the characters `$` (dollar sign) or `_` (underscore). This allows symbols to contain a reserved keyword as a prefix:
+
+```ion
+// All single symbols, even when prefixed by a keyword
+trueSymbol
+null123
+nanfalse
+```
+
+This means that any value that begins with a non-identifier character can appear immediately after an identifier or one of these reserved keywords:
+
+```ion
+// All valid pairs of two top-level values
+abc-5
+symbol["And a list"]
+anotherOne'andAnother'
+true(story)
+null.float+inf
+```
+
+Typed nulls do not form valid Ion if not followed by a non-identifier character:
+
+```ion
+null.struct5E10    // ERROR: not a valid Ion value - struct5E10 is a bad null decorator
+```
+
+## Numeric stop-characters
+
+Ion text enforces stricter rules on certain "numeric" types. Ion text enforces that integers, real values (decimals and floats, including the special float values `-inf` and `+inf` but excluding `nan`), and timestamps must be followed by one of the following 15 numeric stop-characters: `{}[](),"' \t\n\r\v\f`.
+
+This means that strings, quoted symbols, lobs, structs, lists, and S-expressions can appear immediately after a numeric or timestamp value in a top-level datagram or S-expression:
+
+```ion
+{% raw %}
+// All of this is valid Ion
+123{a: "struct"}456
+-0.5[a, list]
+5.34e9"then a string"
+0xdeadbeef'hello'::world
++inf{{ "<-- also works with -inf" }}
+(1["list"]2000T{a: "struct"}0b010101("sexp")4)
+{% endraw %}
+```
+
+Anything that is not a numeric stop-character appearing immediately after a numeric or timestamp value is a syntax error. This notably includes comments and S-expression operators:
+
+```ion
+123// a comment              // ERROR: single-line comment is not a valid numeric stop
```

Comment:
This is surprising. Comments are explicitly called out in the spec as being equivalent to whitespace, and I believe that all of our implementations treat comments as a valid whitespace for the purposes of terminating a numeric value.